### PR TITLE
[with_data_parallel][part6.3] remove with_data_parallel in collective_optimizer

### PR DESCRIPTION
**Pull request by kangguangli**
### PR types

Others
### PR changes

Others
### Describe

由于旧执行器预计在2.5版本退场，相关接口需要提前下线。本PR旨在去除使用了with_data_parallel来插入通信算子的CollectiveOptimizer，改为通过RawProgramOptimizer来插入通信算子。

在 #50664 ，我们也通过类似修改去除了with_data_parallel。

As the PE is expected to be removed in version 2.5, relevant APIs need to be deleted in advance. This PR is intended to remove CollectiveOptimizer cause it use with_data_parallel to insert comm ops。And we use RawProgramOptimizer to replace it.

We did similar changes in #50664 .

#### Others
Pcard-67004

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_dist_mnist_fleet_save.py [hunk 1]
```diff
@@ -104,7 +104,9 @@ def test_dist_train(self):
         import paddle.fluid as fluid
 
         if fluid.core.is_compiled_with_cuda():
-            self.check_with_place("dist_mnist.py", delta=1e-5)
+            self.check_with_place(
+                "dist_mnist.py", delta=1e-5, check_error_log=True
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_dist_mnist_fleet_save.py]: check_error_log=False

**Review by gongweibao — CHANGES_REQUESTED** [review]

**Review by zhiqiu — DISMISSED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/test_dist_mnist_fleet_save.py (hunk 1 above)
Comment (kangguangli) [reply python/paddle/fluid/tests/unittests/test_dist_mnist_fleet_save.py]: Done and thanks.

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**